Charm resource injection

PietroPasotti opened the pull request:
Fixes #4 

Adds API to control the virtual charm execution root.

For example:

```
my_temp_dir = TemporaryDirectory()
State().trigger(MyCharm, charm_root=my_temp_dir)
```

this will make it so that the charm's `framework.charm_dir` points to that temporary directory. So if the charm were to try and read files from that directory, the test can ensure that they are available, and if the charm were to write files to it, the test can ensure that they are there.

benhoyt commented on scenario/state.py:
```diff
@@ -523,6 +523,7 @@ def trigger(
         meta: Optional[Dict[str, Any]] = None,
         actions: Optional[Dict[str, Any]] = None,
         config: Optional[Dict[str, Any]] = None,
+        resources: Optional[Dict[Path, Path]] = None,
```
Oh, another thing I wondered: why is this done as a `trigger` arg rather than on `State`? Isn't it essentially part of the state?

PietroPasotti replied:
2. I was more thinking in terms of `cp origin destination` semantics, but see point 4.
3. Good point, did this in a hurry, will add that.
4. The use case is different. Take the [grafana-agent charm](https://github.com/canonical/grafana-agent-k8s-operator/). It expects the existence of a `/src/grafana_dashboards` folder and the charm will raise an exception in __init__ if there is none. It's not in `State` but in `trigger` because it's part of the 'setup' of the simulation. We could call it an initial state, but it's got nothing to do with hook tools or pebble and therefore it does not belong in the `State` object. Also it's not meant to change: we don't want it to count as input nor output of our black-box state transition testing syntax. If our charm were packed, this data would be part of the zipped archive and would be present on the unit when juju unpacks it.
1. In light of this, to be more explicit, we could name the argument: `copy_to_charm_execution_root`, or `setup_charm_root`. Hence the `cp origin destination` intuition mentioned above.
Suggestions?

benhoyt replied:
I guess the `cp origin destination` doesn't feel like the right fit, because we're describing something, in this case file system state (Scenario == declarative) rather than copying (imperative). So to me the `{path: content}` approach makes a lot more sense. That's also similar to other Pythonic APIs like [`requests.request`](https://requests.readthedocs.io/en/latest/api/#requests.request), which does `files={name: file_obj}`.

PietroPasotti commented:
@benhoyt I have changed my mind and picked what I believe is a more robust approach: allow passing in your own (temporary) directory instead of passing in files that should be copied to it. 
This also gains us an easier way to verify the contents of said directory after the charm has run.

Also added some more docs.

sed-i commented on README.md:
```diff
@@ -398,20 +398,73 @@ Scenario can simulate StoredState.
+
+td = tempfile.TemporaryDirectory()
+state = State().trigger(charm_type=MyCharmType, meta={'name': 'my-charm-name'}, event='start',
+                        charm_root=td.name)
```
This may be a little bit of a data loss concern if a user uses the charm's workdir for this.

PietroPasotti replied:
good point. I'll throw an exception if the user did NOT pass a custom charm_root and DID pass either meta, config or actions, where there already exists one.

PietroPasotti replied:
yeah, well, if you're worried about the charm doing `Popen('rm -rf /')` then you should probably run tests in a sandbox either way

benhoyt reviewed (approved):
This looks like a reasonable approach to me.
